usb/esp_tinyusb/tinyusb_net.c

```diff
@@ -0,0 +1,173 @@
+/*
+ * SPDX-FileCopyrightText: 2023 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include "freertos/FreeRTOS.h"
+#include "freertos/event_groups.h"
+#include "tinyusb_net.h"
+#include "descriptors_control.h"
+#include "usb_descriptors.h"
+#include "device/usbd_pvt.h"
+#include "esp_check.h"
+
+#define MAC_ADDR_LEN 6
+
+typedef struct packet {
+    void *buffer;
+    void *buff_free_arg;
+    uint16_t len;
+    esp_err_t result;
+} packet_t;
+
+struct tinyusb_net_handle {
+    bool initialized;
+    SemaphoreHandle_t buffer_sema;
+    EventGroupHandle_t  tx_flags;
+    tusb_net_rx_cb_t    rx_cb;
+    tusb_net_free_tx_cb_t tx_buff_free_cb;
+    tusb_net_init_cb_t init_cb;
+    char mac_str[2 * MAC_ADDR_LEN + 1];
+    void *ctx;
+    packet_t *packet_to_send;
+};
+
+const static int TX_FINISHED_BIT = BIT0;
+static struct tinyusb_net_handle s_net_obj = { };
+static const char *TAG = "tusb_net";
+
+static void do_send_sync(void *ctx)
+{
+    (void) ctx;
+    if (xSemaphoreTake(s_net_obj.buffer_sema, 0) != pdTRUE || s_net_obj.packet_to_send == NULL) {
+        return;
+    }
+
+    packet_t *packet = s_net_obj.packet_to_send;
+    if (tud_network_can_xmit(packet->len)) {
+        tud_network_xmit(packet, packet->len);
+        packet->result = ESP_OK;
+    } else {
+        packet->result = ESP_FAIL;
+    }
+    xSemaphoreGive(s_net_obj.buffer_sema);
+    xEventGroupSetBits(s_net_obj.tx_flags, TX_FINISHED_BIT);
+}
+
+static void do_send_async(void *ctx)
+{
+    packet_t *packet = ctx;
+    if (tud_network_can_xmit(packet->len)) {
```

Comment:
How can we signal failure of this call to the user?

Reply:
This is an asynchronous API, so the packet would be just dropped.
I've added an error message, but that's probably all we can do I think.

Maybe we could add Tx result argument to this `tx_buff_free()` callback, WYT?

Reply:
> Maybe we could add Tx result argument to this tx_buff_free() callback, WYT?

This sounds good to me. But if it is OK to drop packages silently, then even the current solution is OK.

Nitpick: would `tx_done_cb` be a better name for this callback?

Reply:
> if it is OK to drop packages silently, then even the current solution is OK.

I'd prefer this option TBH. It's already described here, that the async mode should be used with consideration (and now with the error message, developers can easily adjust task priorities to reduce this kind of packet drop)

https://github.com/espressif/idf-extra-components/blob/d3a6806d194ce934871866f0f62c478c98e4e931/usb/esp_tinyusb/include/tinyusb_net.h#L83-L84

> Nitpick: would tx_done_cb be a better name for this callback?

`tx_done_cb()` wouldn't reflect the purpose of this callback IMO, as this is called exactly the moment when we *copy* the data from *our* buffer to the internal tinyUSB buffer. so the exact moment we're able to deallocate the buffer.
We could also add a similar callback for transmit completion (and maybe add a completion status), but I would probably like to keep this simple (adding a new callback isn't a breaking change and can be implemented later)